Add separate endpoint for create and invite participant group methods…

yuanchen233 opened the pull request:
… to RecruitmentService

Also marks `InviteNewParticipantGroup` Endpoint and related method as deprecated, and replace their usage with new create and invite method respectively.

----------
- I did not include `connectedDevicePreregistration` here, as I think it should be a separate issue/commit.
- I did not include 'add `UpdateParticipantGroup` endpoint' in this PR, to keep thing clean here.

Qs:
- Should test cases related to `inviteNewParticipant` be removed? as its functionalities are now in create and invite, respectively and we have coverage for them. So some suppress can be removed. ( Kotlin compiler fails it even the deprecation level is warning )


---
TODO(not for this PR):
1. Update 'stopParticipantGroup' parameter in the order from the most relevant one to secondary ones.
2. Create an issue to keep all creation endpoint consistent: let user pass IDs.
3. Create new endpoint `UpdateParticipantGroup`
4. Rename `StagedParticipantGroup `

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,12 +74,51 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+     * and an optional [name] representing this group, but do not yet send out invitations.
+     * This is used to create a group of participants which can be deployed at a later time.
+     *
+     * In case a group with the same participants has already been deployed and is still running (not stopped),
```
This feels weird. Now that create/invite is separated, do we still need this same behaviour? As long as the same parameters are provided, can't we always return the status? Since there is no "invite" expectation now, I think this original restriction no longer makes sense. 

If the caller passes the groupID, the contract is even clearer. 

And, I guess we could allow creating groups after the study has stopped, they just can't be deployed.

Whathecode replied:
And, considering the idempotent implementation, do we actually really want that?

Take a look [at how this is done for](https://github.com/carp-dk/carp.core-kotlin/blob/develop/carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/application/DeploymentService.kt#L67) `createStudyDeployment`.

I'm leaning towards just being consistent instead of trying to be fancy, unless there is some clear recommendation to have idempotent "create" operations (we should do some pro/con reading). If it is preferred, we should likely create a task to make all create operations in CARP idempotent.

Whathecode replied:
I think [the main reason for this being idempotent in the past](https://github.com/carp-dk/carp.core-kotlin/issues/125#issue-613229634) is resolved by both splitting these calls, and letting the caller pass the ID. 

Given the ID (if we add it), the create call will fail a second time around. And, the invite call simply shouldn't send out invites again. There is no need to verify parameters to assess whether it is a new or existing group.

It's both consistent and simpler.

yuanchen233 replied:
I agree that this behavior is no longer needed.

> I'm leaning towards just being consistent instead of trying to be fancy, unless there is some clear recommendation to have idempotent "create" operations 

In the current flow of creating a new study deployment (in CAWS at least):  `createStudyDeployment` is called by `RecruitmentService.inviteNewParticipantGroup`. Therefore a groupID is provided by the recruitment subsystem, which keeps all `UUID` generation in the backend.

We also have example of [createStudy](https://github.com/carp-dk/carp.core-kotlin/blob/50ef1471e20639837c4b1a1fdb95d79748abd0da/carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/StudyServiceHost.kt#L27) which generate a new ID for the study.

Side note: what happens if `CreateStudyDeployment` called directly, would It be a deployment not tracked by any study?

----
 Actually, I believe allowing the caller to pass the groupID helps maintain isolation between subsystems. This leaves the potential of creating study deployments first, and add them to a study later; or separate `Study` subsystem and `StudyDeployment` subsystem completely.

Whathecode replied:
> We also have example of [createStudy](https://github.com/carp-dk/carp.core-kotlin/blob/50ef1471e20639837c4b1a1fdb95d79748abd0da/carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/StudyServiceHost.kt#L27) which generate a new ID for the study.

If we ever need to refactor that, it would make sense to align and let the caller pass an ID. All in all, I seem to recall that is best practice. So, maybe something you can add as an issue to the backlog. :) (not urgent though, I bet)

> Side note: what happens if CreateStudyDeployment called directly, would It be a deployment not tracked by any study?

Yes; the whole point is you can have deployments without studies, or different study subsystems could decide how to manage their deployments. Or, you can use a hardcoded deployment for a singular app, etc ...

yuanchen233 replied:
I remembered keeping that behaviour so the deprecated inviteNewParticipantGroup would keep working; that logic now lives inside the legacy method. 

I also added a guard in createParticipantGroup so if a group already exists under that groupId but its config doesn’t match, we let users know a new group wasn’t created.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,12 +74,51 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+    ): ParticipantGroupStatus
+
+    /**
+     * Invite the participant group with the specified [groupId] (equivalent to the studyDeploymentId)
```
Isn't the equivalence between deployment ID and groupID an implementation detail? I.e., not something to document/make part of the contract.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,12 +74,51 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+     *
+     * @throws IllegalArgumentException when:
+     *  - a study with [studyId] or participant group with [groupId] does not exist.
+     *  - participant role assignment is empty
```
Wrong documentation.

And, we need to document what happens when the group has already been invited (like the old behavior).

yuanchen233 replied:
I'm not following this, should empty role assignment be `IllegalStateException`?

I'm throwing `IllegalStateException` when group has already been deployed. Should I keep the old begavior to return running group's status here?

Whathecode replied:
> any of the participant roles specified in [group] does not exist

Can this ever happen? This is specified when you create the group (or later when you update it). So, a group with invalid participant roles can't exist at this point in time. That's an invariant protected by this service.

> not all necessary participant roles part of the study have been assigned a participant

That one I guess can still happen. And, can only be resolved once you allow the user to update (but it's fine that isn't part of this commit).

I guess it also covers "participant role assignment is empty" which is redundant then?

yuanchen233 replied:
Got it, thanks!

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -124,14 +125,53 @@ class RecruitmentServiceHost(
      *  - not all necessary participant roles part of the study have been assigned a participant
      * @throws IllegalStateException when the study is not yet ready for deployment.
      */
+    @Deprecated(
```
Are you certain this is needed on both interface and concrete class?

yuanchen233 replied:
Gradle and our detekt are okay with it.

I added it because IntelliJ warns about overriding a deprecated member but not mark it as deprecated.

Whathecode reviewed (commented):
Only reviewed the interface for now.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -124,14 +125,53 @@ class RecruitmentServiceHost(
+        }
+
+        // Create study deployment, which sends out invitations.
+        // TODO: If the repository gets updated but `createStudyDeployment` fails, state will be inconsistent.
```
The repository update and `createStudyDeployment` are no longer visible in this method. Probably more legible to write "if `createParticipantGroup` succeeds, but `inviteParticipantGroup` fails, repository state will be inconsistent.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -145,17 +185,39 @@ class RecruitmentServiceHost(
+        val (recruitment, participantGroup) = getRecruitmentAndGroupOrThrow( studyId, groupId )
+        check( !participantGroup.isDeployed ) { "Participant group has already been deployed." }
+
+        val (protocol, invitations) = recruitment.createInvitations( participantGroup.roleAssignments )
```
The original code mentioned:
>         //  This should use eventual consistency: https://github.com/imotions/carp.core-kotlin/issues/295

I believe that is still relevant, since there is a call to `deploymentService` here. That call may succeed, but subsequent repo update may fail.

Not that this has to be fixed in this PR, but the comment/task should remain.

yuanchen233 replied:
This comment were left untouched in old `inviteNewParticipantGroup` method, I will move it down to new endpoint.

Whathecode replied:
Yes, but it is also relevant for the new endpoint. :)

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -169,12 +231,15 @@ class RecruitmentServiceHost(
+            else deploymentService.getStudyDeploymentStatusList( deployedStudyDeploymentIds )
 
-        return studyDeploymentStatusList.map { recruitment.getParticipantGroupStatus( it ) }
+        return stagedParticipantGroups + studyDeploymentStatusList.map { recruitment.getParticipantGroupStatus( it ) }
```
I was wondering why you updated this. But, I see it is because you also return _staged_ participants now. Make sure to update the documentation (it mentiones "all deployed participant groups"). I guess, since before you couldn't have just "staged" participants, we can still consider this fully backwards compatible.

You can make this code more readable by not intermingling getting the status of staged groups with that of deployed groups. They could be two separate "code blocks" with a comment above.

It's also not great that the status of one is created _within_ `recruitment`, and the other as a private method here. Can't both be encapsulated in `recruitment`? Maybe even the same method but with a nullable deployment status (meaning, not deployed).

yuanchen233 replied:
This change could probably be a separate commit — my main goal was just to make it possible to get info about `StagedParticipantGroup` after creating them.

It was pulled out into a private method because:
1. The multi-layer nested `.map` calls didn’t look clean; and  
2. Building a lookup table seemed potentially more efficient.

I was trying to find a good place for the method that builds `ParticipantGroupStatus.Staged`.  
Since it depends on `recruitment`, it couldn’t live inside `StagedParticipantGroup`.  
Now that you mention it, `recruitment` does seem like the right place.

I’m thinking of adding it as a separate method in `recruitment`.  
`DeployedGroups` needs `deploymentStatus`, while `StagedParticipantGroup` only needs `groupID`;  
but since `getParticipantGroupStatus` returns a `ParticipantGroupStatus`, would method overloading make sense here?

> 💡 By `StagedParticipantGroups` I mean non-deployed groups — maybe we should consider renaming it for clarity. I think you also mentioned rename it in an earlier issue to `ParticipantGroup`, but it's taken by deployment subsystem already..

Whathecode reviewed (commented):
Reviewed `RecruitmentServiceHost`. Primarily, we should try to come to a decision on whether or not we want idempotency and then pursue something consistent.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,14 +74,53 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+     *  - any of the participant roles specified in [group] does not exist
+     * @throws IllegalStateException when the study is not yet ready for deployment.
+     */
+    suspend fun createParticipantGroup(
```
We also need to add this endpoint to the docs: https://github.com/carp-dk/carp.core-kotlin/blob/develop/docs/carp-studies.md
Maybe I'll take that on me, as the state diagram should probably also be updated. But, you could already update the text, though. :)

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,14 +74,53 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+     * and an optional [name] representing this group, but do not yet send out invitations.
+     * This is used to create a group of participants which can be deployed at a later time.
+     *
+     * When [groupId] identifies an existing group, the latest status for this group is returned.
```
I was actually thinking of just throwing here, consistent with `DeploymentService`. https://github.com/carp-dk/carp.core-kotlin/blob/develop/carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/application/DeploymentService.kt#L60

Or, consistent with `inviteParticipantGroup` in this service which throws `IllegalStateException`.

Do you see a benefit in doing this instead?

yuanchen233 replied:
The only potential benefit would be that the user can get the status of the group back using the same method call. However, since `GetParticipantGroupStatusList` now includes staged groups, it should be the recommended approach.

I’ll go with `IllegalStateException`, as the provided argument is valid, but "an object is in an invalid state for that operation.” We should agreed on which one to use and make it consistent across the system.

---
Update: I ended up with `IllegalArgumentException`, as "the caller supplied parameters that violate the method’s precondition".

I recall that I have seen people arguing about the usage between `IllegalArgumentException` vs `NullPointerException`, when the provided non-nullable argument is `null` a long time ago, now here I am :)

Whathecode replied:
It is indeed an illegal argument. 😉 The object which is supposed to safeguard state doesn't exist yet, so that'd be odd.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -124,57 +125,153 @@ class RecruitmentServiceHost(
+        // Create study deployment, which sends out invitations.
+        // TODO: If createParticipantGroup succeeds, but inviteParticipantGroup fails, repository state will be inconsistent.
+        //  This should use eventual consistency: https://github.com/imotions/carp.core-kotlin/issues/295
+
```
Remove newline. This comment belongs to the `return` that follows immediately.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -124,57 +125,153 @@ class RecruitmentServiceHost(
+            if ( !existingGroup.isDeployed )
+            {
+                // A staged group should still go through the deprecated flow and trigger invitations.
+                return inviteParticipantGroup( studyId, existingId )
```
I think I can see what you tried to do here, but, this wasn't part of the old `inviteNewParticipantGroup` behavior, and neither is this inviting a _new_ participant group as per the contract of this method.

Instead, it's supporting a call to the new `createParticipantGroup` followed by a call to `inviteNewParticipantGroup`, as far as I can follow.

It's also not addressing the main comment above:
>  Support the deprecated API by reusing groups that were created through repeated calls with the same configuration ...

On a more personal note, the multiple returns hamper readability. They are great for clear early outs (such as the previous "existing group" edge case), but now there are 4 returns, two of which to the same method, with space between them. The behavior is less clear at a glance.

yuanchen233 replied:
Yep, I was trying to reuse some code, as the plan was to separate creation and invitation to participant groups, but the behaviour of those endpoints are changed now. I went to far on this route :(

I'll revert this endpoint/method and leave it as is.

Whathecode replied:
It's deprecated code; no harm in not maintaining it. ;)

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,14 +74,53 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+     */
+    suspend fun createParticipantGroup(
+        studyId: UUID,
+        groupId: UUID,
```
Having this as the first parameter would be more consistent.

And, having `groupId` and `group` be followed by `studyId` could reduce the change of passing wrong IDs (`groupId` to `studyId` and vice-versa).

yuanchen233 replied:
Should this be applied to `inviteParticipantGroup` and `stopParticipantGroup` as well? Since they also take `studyId` and `groupId`.

Whathecode replied:
Well spotted! it is indeed a good convention to have endpoints operating on an object(s) take the identifier of said object as the first parameter. Even more important, to be consistent. Go from the most important/relevant parameters to secondary ones; it helps intuition in the API.

yuanchen233 replied:
I updated `inviteParticipantGroup` regarding this change, I'll update `stopParticipantGroup` in a separate commit later on to keep this commit focused.

Whathecode replied:
We could also consider removing `studyId` entirely. If groups have unique IDs, `studyID` shouldn't be needed after creation. 🤔 But yeah, separate change, maybe even other PR.

yuanchen233 replied:
Yep, we could. I kept `studyId` here since `getRecruitmentOrThrow` requires it, we currently don't keep track of `studyId` in participant groups or deployments, and loop through all recruitments are inefficient.

We could keep `studyId` inside each `StagedParticipantGroup` to support this.

Whathecode replied:
For optimization it is maybe a bit premature. Unless you run massive studies you won't feel this, and then the first limitation will likely be persistence instead. 😉

But, from a design perspective I guess there is no more reason to do this differently. I think I just took some shortcuts back when I designed this.

yuanchen233 replied:
I added a new helper `getRecruitmentAndGroupByGroupIdOrThrow` to do this, and `stopParticipantGroup` can reuse it later.
(better name for this are very much welcome

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -124,57 +125,153 @@ class RecruitmentServiceHost(
 
-        // Create study deployment, which sends out invitations.
         // TODO: If the repository gets updated but `createStudyDeployment` fails, state will be inconsistent.
         //  This should use eventual consistency: https://github.com/imotions/carp.core-kotlin/issues/295
```
If `createStudyDeployment` fails, this method would fail. The problem is different now: `createStudyDeployment` may succeed, but `updateRecruitment` may fail.

This comment applies to the `"Mark participant group as deployed" code block now.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -124,57 +125,153 @@ class RecruitmentServiceHost(
+            else deploymentService.getStudyDeploymentStatusList( deployedStudyDeploymentIds )
 
-        return studyDeploymentStatusList.map { recruitment.getParticipantGroupStatus( it ) }
+        return stagedParticipantGroups + studyDeploymentStatusList.map { recruitment.getParticipantGroupStatus( it ) }
```
For consistency, move the `.map { recruitment.getParticipantGroupStatus( it ) }` to the "Get status list for deployed participant groups" block.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -124,57 +125,153 @@ class RecruitmentServiceHost(
+            .values
+            .map { recruitment.getParticipantGroupStatus( it ) }
+
         // Get study deployment status list.
```
This is getting the IDs, ... not the status list. And, this code block is no longer meaningful. Just inline it with "Get status list for deployed participant groups" below.

Whathecode reviewed (changes requested):
Still reviewing, but here are some further comments. :)

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,14 +74,49 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+     *  - a study with [studyId] does not exist
+     *  - an existing participant group with [groupId] already exists
+     *  - any of the participant roles specified in [group] does not exist
+     * @throws IllegalStateException when the study is not yet ready for deployment.
```
Re-reading this again, I think this exception warrants a bit more explanation. I would add:

> As long as no final study protocol is locked in for the study, a participant group can't be created since participant roles to which participants need to be assigned are unknown.

Don't forget to also update `RecruitmentServiceHost`. :)

That's at least where the current limitation comes from. But, that does seem limiting and perhaps a useful area to improve functionality for later on.

In the main use case there is only a single participant and participant roles are irrelevant. That's why `AssignTo.All` exists. If group only contains `AssignTo.All`, that should be a valid configuration for any study protocol, even if it changes after.

And/or, maybe it is fine to only validate the assignments in context of the _current_ protocol, make the assignments nullable, and simply invalidate existing role assignments when a new protocol is set.

Does this make sense? If so, we should likely spin off a new feature/discussion issue for this.

yuanchen233 replied:
Yes, this feature would decouple study setup from inviting participants. So as a researcher, I can recruit participant before finalize technical details "protocol".

> In the main use case there is only a single participant and participant roles are irrelevant. That's why AssignTo.All exists. If group only contains AssignTo.All, that should be a valid configuration for any study protocol, even if it changes after.

Since protocol also defines the number of participant, `AssignTo.All` could be misused here. But yep, the discussion should be hold in a separate issue.

yuanchen233 replied:
> Re-reading this again, I think this exception warrants a bit more explanation. I would add:
> As long as no final study protocol is locked in for the study, a participant group can't be created since participant roles to which participants need to be assigned are unknown.

Should this also be added to `Recruitment.addParticipantGroup`'s documentation, or the error message?

Whathecode replied:
I'm afraid you squashed fixups for this into the wrong commit (the second, instead of the first).

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,14 +74,49 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+     * Invite the participant group with the specified [groupId] to start participating in its study.
+     *
+     * @throws IllegalArgumentException when:
+     *  - the participant group with [groupId] does not exist.
```
Minor: drop punctuation at the end. Seemingly it's not used in any of the other bullet lists.

Whathecode replied:
I'm afraid you squashed fixups for this into the wrong commit (the second, instead of the first).

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -204,4 +277,17 @@ class RecruitmentServiceHost(
         participantRepository.getRecruitment( studyId )
             ?: throw IllegalArgumentException( "Study with ID \"$studyId\" not found." )
+
+    private suspend fun getRecruitmentAndGroupByGroupIdOrThrow(
```
I think you can safely drop "ByGroupId" as that is already documented by the parameter name.

Whathecode replied:
I'm afraid you squashed fixups for this into the wrong commit (the second, instead of the first).

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/users/ParticipantRepository.kt:
```diff
@@ -20,6 +20,11 @@ interface ParticipantRepository
+    /**
+     * Returns all stored [Recruitment]s.
+     */
+    suspend fun getRecruitments(): List<Recruitment>
```
This is a very bad idea. It's a full table dump and will inevitably get slower and slower.

I guess you added this to be able to drop the `studyId` parameter. But, the way the repository pattern in DDD works is that you are supposed to expose the _queries_ you need from the domain layer in the interface (the use cases), instead of the very typical simplistic CRUD operations and end up writing query code in the layer above.

So, instead what we likely need is a `getRecruitmentWithParticipantGroup( groupId )`. This can be optimized by whatever underlying DB technology and doesn't pull in unnecessary data.

Whathecode replied:
I'm afraid you squashed fixups for this into the wrong commit (the second, instead of the first).

Whathecode reviewed (changes requested):
Some minor comments, but then I ran into the `getRecruitments()` method on the repository, which is a bigger problem and I'll already post now so you can fix it while I continue reviewing later on. :)

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentService.kt:
```diff
@@ -74,14 +74,51 @@ interface RecruitmentService : ApplicationService<RecruitmentService, Recruitmen
+     * @throws IllegalStateException when the study is not yet ready for deployment;
+     *  As long as no final study protocol is locked in for the study, a participant group can't be created
+     *  since participant roles to which participants need to be assigned are unknown.
+     */
```
If I'm not mistaken, you shouldn't have text underneath `@throws`. The documentation is first, and then additional stuff. So this should be part of the "main" documentation.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/users/ParticipantRepository.kt:
```diff
@@ -20,6 +20,14 @@ interface ParticipantRepository
+     */
+    suspend fun getRecruitmentWithParticipantGroup(
+        groupId: UUID
+    ): Pair<Recruitment, StagedParticipantGroup>?
```
I'm not sure returning `StagedParticipantGroup` in addition to `Recruitment` instead of just `Recruitment` makes sense. It feels inconsistent with the rest of this repository which focuses on `Recruitment`. Also, with the `groupId` retrieving the `StagedParticipantGroup` is a single line of code (and a map lookup) for the caller. So the win vs. added complexity here is extremely minimal.

At the very least, the documentation is confusing. It now reads like it returns the `StagedParticipantGroup` chich contains the specified `groupId`. :)

I may have set you on the wrong track with my suggested naming "WithParticipantGroup". But, I what I meant by that was that the recruitment has that participant group. :) Maybe `getRecruitmentContainingParticipantGroup` or `getRecruitmentForParticipantGroup` is better? Whathever name you land on, it should be clearer from the documentation. :)

yuanchen233 replied:
The name is fine, we also have `getRecruitmentWithGroupOrThrow` and keep name aligned makes sense for consistency.

I kept the signature this way because repeating a `recruitment[groupId]` null check after repository already validated it felt redundant; while since repository implementations might be rewritten in implementations later, being a bit defensive here could pay off later. I couldn't make up my mind so I returned both 😶

I also removed the pair return for the helper method in `recruitmentServiceHost` to keep code in a similar style; and overloaded `getRecruitmentWithGroupOrThrow` name for this method since the parameter is different and that's the most suitable name.

yuanchen233 committed:
Retrieve Recruitment containing a specified group ID from repo

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -160,21 +167,91 @@ class RecruitmentServiceHost(
+
+        require( recruitment.participantGroups[ groupId ] == null ) {
+            "Participant group with ID \"$groupId\" already exists."
+        }
```
This constraint should be enforced by `Recruitment.addParticipantGroup`, not the application service.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -160,21 +168,86 @@ class RecruitmentServiceHost(
+    ): ParticipantGroupStatus
+    {
+        val recruitment = getRecruitmentOrThrow( studyId )
+        require( participantRepository.getRecruitmentWithParticipantGroup( groupId ) == null )
+            { "Participant group with ID \"$groupId\" already exists." }
```
@yuanchen233 Note that I changed this to check for _any_ `groupId`, instead of an existing `groupId` within the `recruitment`.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/application/RecruitmentServiceHost.kt:
```diff
@@ -160,21 +168,86 @@ class RecruitmentServiceHost(
+    */
+    override suspend fun inviteParticipantGroup( groupId: UUID ): ParticipantGroupStatus
+    {
+        val (recruitment, participantGroup) = getRecruitmentWithGroupOrThrow( groupId )
```
@yuanchen233 I rolled this back to return the pair, because in this context it does serve a purpose. And, it's also private so less impactful. For the repository interface, I just didn't want to cause confusion as to the group being contained _in_ the recruitment.

Whathecode commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/users/Recruitment.kt:
```diff
@@ -216,15 +216,34 @@ class Recruitment( val studyId: UUID, id: UUID = UUID.randomUUID(), createdOn: I
+     *
+     * @throws IllegalArgumentException when the specified [group] is not part of this recruitment.
+     */
+    fun getParticipantGroupStatus( group: StagedParticipantGroup ): ParticipantGroupStatus.Staged
```
There are some design issues here:
- The caller can pass _any_ group (an actual deployed group, or even from an entirely different recruitment) and that would go unnoticed.
- The logic of whether to call this method or the one with the `StudyDeploymentStatus` parameter really shouldn't live in the application service.

I think this calls for an external "domain service" (in DDD terminology) which aggregates the two together (study deployments and participant groups) and is called from the application service. So both your new `getParticipantGroupStatus`, but also the previous one I added, should live elsewhere (and probably be merged). Maybe a `ParticipantGroupStatusService`. That would make a logical preceding commit.

`participantGroups` also still documents "deployment IDs", but that must group IDs, now. Slightly related, the naming/documentation of `StagedParticipantGroup` doesn't feel spot on. It starts out staged, for sure, but it seems a bit of a misnomer once deployed. :) Since we plan to allow to update this even _after_ deployment (name change, for example), we should maybe start thinking about a better name 🤔. I can't think of any atm.

yuanchen233 replied:
> The caller can pass any group (an actual deployed group, or even from an entirely different recruitment) and that would go unnoticed.

We throw IllegalArgumentException for this, as we `requireNotNull( _participantGroups[ deploymentId ]`. For deployed groups, it could be a simple `isDeployed` check. I do get the point for a "domain service" here, after some reading on DDD :)

Hi @Whathecode, I am not sure if I'm on the right track here, so I pushed what I have to 'fail early'.

I had different approaches:

1. Passing the current `recruitment` to `ParticipantGroupStatusService.getParticipantGroupStatus()`, so it can check membership, and have access to `participants` by moving `getParticipantsFor` here; and also it can check if a group `isDeployed`.
2. Also making `ParticipantGroupStatusService` depend on `StudyDeploymentService`; then `getParticipantGroupStatus` only requires the `groupId`, and `StudyDeploymentStatus` can be fetched.
3. Current implementation, I am not passing current `recruitment` as it against the purpose of this new service: arguably it should be free of recruitment. I didn't go with approach 2 either because domain service should not depend on another application service, per my understanding so far.

Therefore, as current implementation it looks more like a `factory` rather than `service` as it only builds group status from provided arguments.

====
For names, I am thinking `recruitmentGroup`, `recruitmentEnrollment`, or `EnrollmentGroup` (or Enrolment for British English).

I don't want to get confused with `Deployment.ParticipantGroup`, I'll ask the team for more ideas :wink:

Whathecode replied:
If we consider studies and deployments truly separate domains, DDD would argue there is no harm overloading the term, and the term having a different meaning depending on context: "deployment" participant groups vs. "study" participant groups. :)

yuanchen233 replied:
We could also do that, and referring them with 'studies' and `deployment` prefix, it's more natural as it corresponds to `ParticipantGroupStatus`; however, `Staged` state feel more like 'Configured' or 'Planned' tho.

Whathecode replied:
Finally looked at this; sorry for the slow responses. :(

I'm thinking a static `ParticipantGroupStatusService` (class with companion object in Kotlin), and not a class with interface. That can live in the domain layer.

Then I am considering a: `ParticipantGroupStatusService.forGroup( group: StagedParticipantGroup, getDeploymentStatus: (UUID) -> StudyDeploymentStatus)` call. The service can conditionally call `getDeploymentStatus` based on `isDeployed`. If you already have the status, you can just inject a lambda returning that (on create). When you don't, you can inject a call to `deploymentService`.

 I see no need to DI this into `RecruitmentServiceHost`, as this opens up the possibility of injecting a different `DeploymentService` into `RecruitmentServiceHost` than that used by calls to `ParticipantGroupStatusService`, which always seems wrong to me. I think information hiding still wins out here. The only benefit of injecting it would be that you can test the implementation does calls to it where expected. But, that is testing the _implementation_ and not the desired _behavior_.

And, to support `getParticipantGroupStatusList` the method should generalize to `forGroups` instead, so that we can benefit from the single aggregate `getStudyDeploymentStatusList` call instead of multiple calls.

Lastly, for review purposes, I think it could make sense to have an initial commit introducing this service and replacing the existing implementation, before introducing the new `create/invite` endpoints.

yuanchen233 replied:
End of years are just busy times..Thanks for the details!
I'm also passing the `participants` to make this service free from `recruitment`.

Whathecode replied:
Looking at it closer now, by injecting the `getDeploymentStatusList` delegate, _maybe_ there is no real problem anymore keeping this functionality in `Recruitment`. 

I started adding commits on top of yours in that direction. If you don't see problems with this, feel free to squash these into your original commit introducing `ParticipantGroupStatusService` (and of course update the commit message), and continue on the same path.

I now see this breaks a detekt test since the max number of methods in `Recruitment` has been exceeded, which _could_ warrant maybe considering `ParticipantGroupStatusService`. :) It can easily be moved, or we could suppress the detekt warning. But, more importantly, I don't see the real problem you had with making this functionality free from `Recruitment` 🤔. And, rather than duplicating logic, I let the singular method call the plural method.

yuanchen233 replied:
My consideration was building `ParticipantGroupStatus` needs `Participants` which is persisted in the recruitments, therefore the caller needs an instance of recruitment to use `ParticipantGroupStatusService`. By using a `getParticipants` delegate like `getDeploymentStatusList`, I don't think it's a concern anymore.

I made `getParticipantsFor` an internal method instead private call it from the host, that adds one more method, changing it back should solve the detekt problem. However, we are still at maximum number of methods.

Looking at the list of methods, can the 3 overloaded `addParticipant` be optimised into one?
I think methods adding participant by email and username is not used currently, the logic is handled in recruitment host.

Whathecode replied:
> the caller needs an instance of recruitment to use ParticipantGroupStatusService.

But why is that a problem? That seems fine to me. 🤔 The delegate for `getDeploymentStatusList` is to prevent coupling from the domain layer to the application service (`RecruitmentService`). The same doesn't apply to retrieving participants through `Recruitment` (which itself is domain layer).

Note: typically there should be _no_ reference from the `domain` to `application` (such as `Recruitment` depending on `StudyDeploymentStatus`). But, I seem to recall that I ended up placing a lot of immutable value types which really belong more to the domain in the application layer because I expose them "as is" through the application layer. That's because of how powerful `kotlinx.serialization` is; I really didn't feel like creating DTOs for each of them. They are serializable as is already within the domain. Then, to make clear what the external surface area is of the application service (and for which schemas need to be created, etc.), I prioritized placing these in `application`.

> I made getParticipantsFor an internal method instead private call it from the host, that adds one more method,

True! I think `ParticipantGroupStatusService` may have been premature. Let's just keep things in `Recruitment` for now. We can always refactor later.

> Looking at the list of methods, can the 3 overloaded addParticipant be optimised into one?

If anything, these warrant suppressing the warning as the overloads really don't add complexity. It's used in tests and removes some boilerplate. As long as `AccountIdentity` is a sealed class hierarchy (and thus not an extension point) I think the overloads make perfect sense to keep.

yuanchen233 replied:
Got it now, thanks, I was confused about where we should use a domain service in this project and the constrains, I found `ParticipantGroupService` and trying to follow it's structure and use case.

yuanchen233 replied:
Added a guard in `stopParticipantGroup` to check if the provided group belongs to the provided study, since we don't use `studyId` to get recruitment anymore.

Whathecode replied:
> I found ParticipantGroupService and trying to follow it's structure and use case.

Interestingly, I see there are also application services injected into the domain service there, which never really seemed to have bothered me. 😅 It may not be 100% DDD, but at least its just interfaces, so maybe not too bad. Injecting the get deployment delegate like we did isn't too different from injecting an interface, except that it applies the interface segregation principle and injects the minimal subset of the interface it needs.🤷🏼‍♂️

I think I may need a refresher on DDD to fully understand how much we should care about this 'violation'.🙂 As long as we don't inject anything concrete it seems fine to me.

Whathecode replied:
> I found ParticipantGroupService and trying to follow it's structure and use case.

Interestingly, I see there are also application services injected into the domain service there, which never really seemed to have bothered me. 😅 It may not be 100% DDD, but at least its just interfaces, so maybe not too bad. Injecting the get deployment delegate like we did isn't too different from injecting an interface, except that it applies the interface segregation principle and injects the minimal subset of the interface it needs.🤷🏼‍♂️

I think I may need a refresher on DDD to fully understand how much we should care about this 'violation'.🙂 As long as we don't inject anything concrete it seems fine to me.

At least [some people online](https://softwareengineering.stackexchange.com/a/456594/15464) seem to think the same and argue it could be fine to place said interfaces in the domain layer. (Not that I think we should start moving stuff around.) It makes sense; it's the request objects that are the 'true' application layer.

yuanchen233 replied:
Sounds good, we have some thing to follow in this project on this matter now.

Whathecode reviewed (changes requested):
The small stuff I simply amended myself. But, I noticed a problem in `Recruitment` which I think warrants the introduction of a `ParticipantGroupStatusService` (or similar if you can come up with a better name :) ).

yuanchen233 committed:
Inject `getDeploymentStatusList` delegate in getParticipantGroupStatus

yuanchen233 committed:
Add separate RecruitmentService endpoints to create/invite participant groups

This replaces the `inviteNewParticipantGroup` endpoint, which is now deprecated.

yuanchen233 commented on docs/carp-studies.md:
```diff
@@ -55,6 +55,8 @@ Allows setting recruitment goals, adding participants to studies, and creating d
-| `inviteNewParticipantGroup` | Create and instantly invite a group of previously added participants to a study. | manage study: `studyId` | |
+| `inviteNewParticipantGroup` | (Deprecated) Create and instantly invite a group of previously added participants to a study. | manage study: `studyId` | |
+| `createParticipantGroup` | Prepare a group of previously added participants for deployment without inviting them yet. | manage study: `studyId` | |
+| `inviteParticipantGroup` | Invite a previously prepared participant group to start participating in the study. | manage study: `studyId` | |
```
For `inviteParticipantGroup`, the recommended claim is manage study, but the endpoint now only requires `groupId`, is this still applicable here? since `getParticipant` and `stopParticipantGroup` will also require only `groupId` in the future.

They are still part of `RecruitmentService`, but not requiring `studyId` as an the argument, so maybe document them separately?

Whathecode replied:
Yes it is. Otherwise people who don't manage the study can invite participants. 🤷🏼‍♂️

yuanchen233 replied:
Yes the claim should be the same, just documentations here.

After removing the `studyId` argument, we can no longer enforce the provided group belongs to the desired study, it's up to the implementations to check this. Therefore, get the `studyId` of the group and check for the claim serves another purpose for the caller. 

It is not a carp.core concern here, just a matter if we want to point this out.

Whathecode replied:
>  it's up to the implementations to check this.

Claims so far have always been an infrastructure concern. These columns are simply suggestions.

But, this does mean that an additional lookup, other than the information in the request (e.g., repo or backend cache), is needed to implement claim-based authorization here, since without that `studyId` is unknown.

yuanchen233 commented on carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/users/ParticipantGroupStatusService.kt:
```diff
@@ -0,0 +1,102 @@
+            require( deploymentStatusIds == deployedIds )
+            {
+                "Deployment status IDs \"$deploymentStatusIds\" don't match requested group IDs \"$deployedIds\"."
+            }
```
There is a change in strictness, `Recruitment.getParticipantGroupStatus` only require each deployment status belongs to current recruitment, now we also check if there are missing/exceeding deployment status compare to participant groups.

Whathecode replied:
When I refactored this I noticed this, and I thought it best to be accepting (so if there are "exceeding" deployment status returned, not fail, but of course we need the status for those requested, and we want to fail early if anything is missing.

yuanchen233 committed:
Add missing request json

Whathecode commented:
@yuanchen233 Full review finally done. :) I made minor changes, which you can see in [my force-pushed changes](https://github.com/carp-dk/carp.core-kotlin/compare/bd1d21ed1a9de42d2773e1ce8f54e6b9257d73f8..df75a5924bcfbd0eae34f8169bf8bb47337a2dd5). Primarily:
- Updated the README code sample based on the readme() code you updated. That is currently a copy/paste still. The actual code is there to ensure the code in the README code compiles. :)
- Spacing here and there.
- More focused tests of `createParticipantGroup` and `getParticipantGroupStatusList`, rather than calling the latter in the former.
- Removed `getParticipantGroupStatusList_returns_staged_groups_as_staged` and instead return/test both deployed and staged group in `getParticipantGroupStatusList_returns_multiple_deployments`.
- Removed TypeScript `can serialize DeployParticipantGroup` test. I think the previous serialization test was there for the more complex `InviteNewParticipantGroup` parameters. So, your update to `CreateParticipantGroup` and testing that was sufficient. 👍
